zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs

```diff
@@ -605,4 +706,119 @@ mod tests {
+            let is_mcopy = cb.query_cell();
+            cb.require_boolean("is_mcopy is bool", is_mcopy.expr());
+            let gadget = MemoryAddrExpandGadget::<F>::construct(cb, is_mcopy.expr());
+
+            ErrOOGMemoryCopyGadgetTestContainer { gadget, is_mcopy }
+        }
+
+        fn assign_gadget_container(
+            &self,
+            witnesses: &[U256],
+            region: &mut CachedRegion<'_, '_, F>,
+        ) -> Result<(), Error> {
+            let [is_mcopy, src_offset, dst_offset, copy_size] =
+                [0, 1, 2, 3].map(|i| witnesses[i].as_u64());
+            self.is_mcopy
+                .assign(region, 0, Value::known(F::from(is_mcopy)))?;
+
+            let src_memory_addr = self.gadget.src_memory_addr.assign(
+                region,
+                0,
+                src_offset.into(),
+                // set length = copy_size + 1 which cause constraint not be satisfied.
+                (copy_size + 1).into(),
+            )?;
+            let dst_memory_addr = self.gadget.dst_memory_addr.assign(
+                region,
+                0,
+                dst_offset.into(),
+                copy_size.into(),
+            )?;
+            // assign memory_expansion_mcopy
+            self.gadget.memory_expansion_mcopy.assign(
+                region,
+                0,
+                0,
+                [src_memory_addr, dst_memory_addr],
+            )?;
+            Ok(())
+        }
+    }
+
+    // test for mcopy case, do constrain: src_address length == dst_address length
+    // so expect specified error.
+    #[test]
+    fn test_invlaid_src_offset_length() {
+        // test is_mcopy = true
+        let witnesses = [0x1, 0x20, 0x30, 0x10].map(U256::from);
+
+        const K: usize = 12;
+        let circuit = UnitTestMathGadgetBaseCircuit::<ErrOOGMemoryCopyGadgetTestContainer<Fr>>::new(
+            K,
+            witnesses.into(),
+        );
+
+        let prover = MockProver::<Fr>::run(K as u32, &circuit, vec![]).unwrap();
+        let result = prover.verify();
+
+        // when mcopy, should encounter constraint error.
+        assert_error_matches(result, "mcopy src_address length == dst_address length");
+    }
+
+    // test for non mcopy case, do not constrain: src_address length == dst_address length
+    // so expect test pass.
+    #[test]
+    fn test_invlaid_src_offset_length_nonmcopy() {
+        // test is_mcopy = false
+        let witnesses = [0x0, 0x20, 0x30, 0x10].map(U256::from);
+
+        const K: usize = 12;
+        let circuit = UnitTestMathGadgetBaseCircuit::<ErrOOGMemoryCopyGadgetTestContainer<Fr>>::new(
+            K,
+            witnesses.into(),
+        );
+
+        let prover = MockProver::<Fr>::run(K as u32, &circuit, vec![]).unwrap();
+        let result = prover.verify();
+        // when not mcopy, should pass.
+        assert!(result.is_ok());
+    }
+
+    fn assert_error_matches(result: Result<(), Vec<VerifyFailure>>, name: &str) {
```

Comment:
we can move this function to a more common file? like test_util.rs?  Can be used later.

Reply:
sure !

Reply:
moved [8a6e54a](https://github.com/scroll-tech/zkevm-circuits/pull/1333/commits/8a6e54a249f71240d87d95ed105eda3229545e43)